Typing improvements for `io` and fixes for `websocketclient`, and a few others

cmidgley opened the pull request:
This is the work I've done to date with `io` focused on making `websocketclient` to be reliable and strongly typed with TypeScript.  It's a fairly substantial set of changes, mostly in the `.d.ts` files but some implementation (especially `websocketclient`).  

Type changes:

- New types for `embedded:io/socket/tcp/tls`, `embedded:network/http/server`, `embedded:network/http/server/options/webpage`, `embedded:network/https/server/options/websocket`, and `embedded:network/websocket/client`
- Improvements to types `embedded:io/socket/listener`, `embedded:io/socket/tcp`,  `embedded:io/socket/udp`, `embedded:network/dns/resolver/udp`, `embedded:network/http/client`, `embedded:provider/builtin`
- Added `Device` (for global variable `device`) into global types to eliminate requirement for import
- Extend the global namespace for `Device` for when imports of extended modules are used (support for `embedded:network/http/server`, `embedded:network/websocket/client`)

Implementation changes:

- Bug fixes/improvements for `websocketclient`:
  - Implement error property via `onError(error: string)` (new feature)
  - Fix issue where control message payloads were not being decoded with the mask (data corruption)
  - Fix an issue where masks were used on both client and server (spec requires masking on client to server only)
  - Fixed issue where fragmented reads used an improper mask (data corruption)
- Fix a bug in `websocket` where using attached sockets would fail (reported in #1428)
- Add common HTTP status codes to `httpserver` on the `HTTP/1.1 <status> <code>` header
- Added options feature to `embedded:network/http/server/options/webpage` to support `.contentType` and `.status`. 

I typed everything based on the current implementation and not `ECMA-419` (many modules are inconsistent with the published 2nd edition of the spec).

cmidgley committed:
Start of io typings work

cmidgley committed:
Adjust namespace for 'config' to `websocketclient/config`

cmidgley committed:
Merge branch 'public' into io-typings

cmidgley committed:
Revert config.js

cmidgley committed:
Reverse webclient manfiest

cmidgley committed:
Revert TLS include

cmidgley committed:
Define Device into global namespace to allow augmentation; have WebSocket extend Device; improvements overall on types

cmidgley committed:
Removed unused import

cmidgley committed:
Fixed module imports to be within declare

cmidgley committed:
More refinements/improvements to tcp/udp/dns/websocket types

cmidgley committed:
More io types, expanded webpage w/status and headers, fixed up 'this' on callbacks

cmidgley committed:
Add status codes to http/1.1 response

cmidgley committed:
Fix bug with attach socket

cmidgley committed:
Further type improvements for websocketclient

cmidgley committed:
Add onError error messages, don't use masks on server (attached sockets), fix bug with masks on fragmented reads, and add mask decode to control messages

cmidgley committed:
Fix bug with reversed client/server on mask

cmidgley committed:
Switch builtin-global to builtin-device to be consistent with other types

cmidgley committed:
Remove unnecessary import

phoddie commented on examples/io/tcp/websocket/WebSocket.js:
I'm not sure about this change. It must fix a problem you see. But, it also would require the caller to set `host`, `port`, and `path` in the options object, which isn't what we want.

cmidgley replied:
Good catch - on review I see I misunderstood what it was doing. I'll fix...

phoddie commented on typings/embedded_network/websocket/client.d.ts:
```diff
@@ -0,0 +1,56 @@
+			dns?: DNSUDPDevice;
+			onReadable?: (this: WebSocketClient, count: number, options: WebSocketClientReadableOptions) => void;
+			onWritable?: (this: WebSocketClient, count: number) => void;
+			onControl?: (this: WebSocketClient, opcode: WebSocketClientOpcode, buffer: Uint8Array) => void; // should this be ArrayBuffer?
```
Yes, `ArrayBuffer` is correct as `onControl` is [specified](https://419.ecma-international.org/#-24-websocket-client-class-pattern-data-format).

cmidgley replied:
Oops.  That was an old comment that I forgot to remove.

cmidgley committed:
Revert prior and apply new fix for attached sockets

cmidgley committed:
Fix type of buffer on onControl

cmidgley commented:
Fixes are in place.  Let me know what else you find..

cmidgley committed:
Small improvements on types
- tls extends tcp instead of full definition
- Add 'undefined' as possible `read` result

phoddie commented on typings/embedded_io/socket/tcp/tls.d.ts:
```diff
@@ -21,12 +21,5 @@ declare module "embedded:io/socket/tcp/tls" {
-		set format(format: string);
-		get format(): string;
-	}
+	export default class TLSSocket extends TCP {}
```
I'm no TypeScript expert, but don't you need to define that the TLSSocket constructor uses `TLSOptions` instead of `TCPOptions` for the options object argument?

phoddie commented on examples/io/listener/httpserver/httpserver.js:
Thank you for filling this in. The change is good. I don't want even more copies of the HTTP status code map in the repo. I'll make a module for those so it can be shared. But that doesn't need to hold this up.

phoddie commented on typings/embedded_io/socket/udp.d.ts:
419 UDP only specifies `udp.read()` - no argument and returns ArrayBuffer augmented with `port` and `address`. It does not specify multicast (though that's definitely on the roadmap)

phoddie commented on typings/embedded_network/http/client.d.ts:
```diff
@@ -20,12 +20,15 @@
+    socket: TCPDevice | TLSDevice;
+    port?: number;
+    host: string;
+    dns: DNSUDPDevice;
```
There's no requirement that this is a DNS using UDP. It could be another resolution transport. But, for the moment this is probably OK?

cmidgley replied:
I'd say we add a generic DNS type when we have more, and change ClientOptions to use that (or just DNSUDPDevice | DNSOtherDevice here).  Either way, it won't be a breaking change as it is just expanding the type.  But using 'any' (or 'undefined') isn't a good alternative as it isn't strongly typed.

cmidgley committed:
Merge latest Moddable public; remove Websocket code fixes (too many upstream changes), fix TLSSocket constructor type, remove paramaterized UDP.read

cmidgley committed:
Import embedded:provider/builtin-device in global.d.ts to ensure Device is defined; remove udp.read(buffer); add TLSSocket.constructor with TLSOptions definition

cmidgley committed:
Adjust for changes to io since initial PR

cmidgley committed:
Minor cleanup for PR

cmidgley commented:
Merged latest code and resolved conflicts.  Reviewed `io` changes since initial PR and applied typing changes as needed.  Removed some no-longer-needed code changes as they are now in base.  Applied changes based on review comments from Peter.  This has been working great for me for the last seven months (on an older version of Moddable) and compiles/passes all my tests with latest code.

Ready for review and possible merging.

phoddie commented:
Finally coming back around to this.... there's so much here that it is a little challenging to review. I'm not asking that you split it up, but it is a bit of an adventure to understand and review.

That said, a lot of it looks good and we should get it merged. The `*-device` stuff is a bit confusing. I think I'm starting to understand it though.
